### ISSUE TITLE
When response code is 204 (or 205), convert is not used, so RxJava2CallAdapterFactory throws the NoSuchElementException

### ISSUE THREAD

**Issue by NeverwinterMoon** [square/retrofit#2867]
I stumbled upon this: https://github.com/square/retrofit/issues/1554 and added a converter because I wanted to handle the empty response body and not pass it as null to Observable. That thread is talking about code 200 and empty body though.

Then I found out that in the current Retrofit version no converter is called at all for 204 and 205:

```
    if (code == 204 || code == 205) {
      rawBody.close();
      return Response.success(null, rawResponse);
    }
```

This forces me to use `Single<Response<SomeModel>>` instead of `Single<SomeModel>` I was hoping to use...

I understand that 204 stands for no content, so it's only logical to pass the response as null. But for my specific case, I would prefer to handle 204 in `onError`, so I hoped to handle the empty body in the converter and just throw a custom error there instead.

Is there any chance that approach will be changed? I am relatively new to Retrofit but I am used to Alamofire on iOS and there it's possible to specify during configuration what is treated as a successful response code:

` .validate(statusCode: 200)`

So every code that is not on that list would be handled as an error.

**Comment by artem-zinnatullin** [comment]
You should be able to add an [OkHttp Interceptor](https://github.com/square/okhttp/wiki/Interceptors) that can convert original response code from `204` to `200` before it gets passed to Retrofit

**Comment by NeverwinterMoon** [comment]
@artem-zinnatullin Aha, thanks for your advice! I just tried the throwing-in-interceptor approach out, and it seems to satisfy my needs.

Nonetheless, I am curious if Retrofit would ever allow to configure custom validation (similar to Alamofire) for response codes or if it's completely against the Retrofit's design ideas.

**Comment by villesinisalo** [comment]
```
    if (code == 204 || code == 205) {
      return Response.success(null, rawResponse);
    }
```

Isn't this violating the `Response` class contract? Reading from `Response.java`:

```
  /** Returns true if {@link #code()} is in the range [200..300). */
  public boolean isSuccessful() {
    return rawResponse.isSuccessful();
  }

  /** The deserialized response body of a {@linkplain #isSuccessful() successful} response. */
  public @Nullable T body() {
    return body;
  }
```

The way I read it, `response.body()` is always non-null if `isSuccesful()` returns `true`. But in fact, if the server returns 204, we get a `null` regardless. 

And because `OkHttpCall` short-circuits the body parsing to `null`, we don't even get a chance to change it to something more sensible with `Converter`s, like for example receiving the empty `ResponseBody` with `BuiltInConverters.responseBodyConverter`:

```
    @DELETE("user/{id}")
    Single<ResponseBody> returns204(@Path("id") int id);
```

So this does *not* work, which is rather unexpcted (empty body is a valid body after all, right?)

**Comment by ZacSweers** [comment]
I'm going to look at trying a PR for better handling of this, but part of the confusion is also that RxJava silently drops the `null` value emitted, and hence why it's a `NoSuchElementException` rather than NPE

https://github.com/ReactiveX/RxJava/blob/0e7b8eaa61f9cac0538ef6a59bfbd0b119b87732/src/main/java/io/reactivex/internal/operators/observable/ObservableSingleSingle.java#L74-L83

**Comment by ZacSweers** [comment]
I've put up a proposal PR in #3003, feedback welcome!

**Comment by jgavazzisp** [comment]
A quick workaround for those who want to start using the 2.6.0, add the below as your first interceptor in OkHttp client. After that, you would be able to call services which returns 204 / 205 as such:
![image](https://user-images.githubusercontent.com/309389/61323072-3a02c700-a863-11e9-9b2f-143c83b48d19.png)

```
import okhttp3.Interceptor
import okhttp3.MediaType
import okhttp3.Response
import okhttp3.ResponseBody
import timber.log.Timber

class EmptyBodyInterceptor : Interceptor {
    override fun intercept(chain: Interceptor.Chain): Response {
        val response = chain.proceed(chain.request())
        if (!response.isSuccessful) {
            return response
        }

        if (response.code() != 204 && response.code() != 205) {
            return response
        }

        if ((response.body()?.contentLength() ?: -1) >= 0) {
            return response
        }

        Timber.d("Rewriting response to contain an empty body.")
        val emptyBody = ResponseBody.create(MediaType.get("text/plain"), "")

        return response
            .newBuilder()
            .code(200)
            .body(emptyBody)
            .build()
    }
}
```

Regards

**Comment by LouisCAD** [comment]
Inspired from the snippet just above, I made a version that doesn't fail when the server responds with a content length of zero, and is lenient if there's actually some content:
```kotlin
import okhttp3.Interceptor
import okhttp3.MediaType
import okhttp3.Response
import okhttp3.ResponseBody

object EmptyBodyInterceptor : Interceptor {

    override fun intercept(chain: Interceptor.Chain): Response {
        val response = chain.proceed(chain.request())
        if (response.isSuccessful.not() || response.code().let { it != 204 && it != 205 }) {
            return response
        }

        if ((response.body()?.contentLength() ?: -1) >= 0) {
            return response.newBuilder().code(200).build()
        }

        val emptyBody = ResponseBody.create(MediaType.get("text/plain"), "")

        return response
            .newBuilder()
            .code(200)
            .body(emptyBody)
            .build()
    }
}
```

**Comment by crisu83** [comment]
Also inspired from the snippets above, here is my version:

```kotlin
import okhttp3.Interceptor
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.Response
import okhttp3.ResponseBody.Companion.toResponseBody
import timber.log.Timber

class NoContentInterceptor : Interceptor {
    
    private val noContentHttpStatusCodes = arrayOf(204, 205) 
    
    override fun intercept(chain: Interceptor.Chain): Response {
        val response = chain.proceed(chain.request())
        
        if (!response.isSuccessful) {
            return response
        }

        if (!noContentHttpStatusCodes.contains(response.code)) {
            return response
        }

        Timber.d("Rewriting response to contain an empty body.")

        return response
            .newBuilder()
            .code(200)
            .body("".toResponseBody("text/plain".toMediaType()))
            .build()
    }
}
```

**Comment by JakeWharton** [comment]
You should still propagate headers

**Comment by RoyRao2333** [comment]
> 

Hi, can you tell me how to properly use ```toMediaType```? I've tried to import the headers and it told me ```Companion``` is not found...

![image](https://user-images.githubusercontent.com/31413093/153365761-fbf60582-bf85-42c5-b56f-d89c6f8c3322.png)

My Retrofit version is 2.9.0. Thanks!

**Comment by RoyRao2333** [comment]
> Also inspired from the snippets above, here is my version:
> 
> ```kotlin
> import okhttp3.Interceptor
> import okhttp3.MediaType.Companion.toMediaType
> import okhttp3.Response
> import okhttp3.ResponseBody.Companion.toResponseBody
> 
> class NoContentInterceptor : Interceptor {
>     
>     private val noContentHttpStatusCodes = arrayOf(204, 205) 
>     
>     override fun intercept(chain: Interceptor.Chain): Response {
>         val response = chain.proceed(chain.request())
>         
>         if (!response.isSuccessful) {
>             return response
>         }
> 
>         if (!noContentHttpStatusCodes.contains(response.code)) {
>             return response
>         }
> 
>         return response
>             .newBuilder()
>             .code(200)
>             .body("".toResponseBody("text/plain".toMediaType()))
>             .build()
>     }
> }
> ```

Hi, can you tell me how to properly use ```toMediaType```? I've tried to import the headers and it told me ```Companion``` is not found...

![image](https://user-images.githubusercontent.com/31413093/153365761-fbf60582-bf85-42c5-b56f-d89c6f8c3322.png)

My Retrofit version is 2.9.0. Thanks!

**Comment by JakeWharton** [comment]
It requires that you are using OkHttp 4.x

**Comment by JakeWharton** [comment]
Yes. Retrofit depend on OkHttp 3.14 and the latest binary compatible release is 4.9.

**Comment by RoyRao2333** [comment]
> Yes. Retrofit depend on OkHttp 3.14 and the latest binary compatible release is 4.9.

Got it! You really saved my day. Thanks again!

**Comment by nikita-bilous** [comment]
I've solved the issue by (in Interceptor):

`return if (response.code == 204 || response.code == 205) {
            response.newBuilder()
                .code(200)
                .body("{}".toResponseBody())
                .build()
        } else {
            response
}`

**Comment by Rajarml** [comment]
This issue hasn't been resolved since the [v2.10.0](https://github.com/square/retrofit/releases/tag/2.10.0) version of Retrofit?
It seems any `EmptyBodyInterceptor` or `NoContentInterceptor` is not required anymore.

**Comment by JakeWharton** [comment]
The behavior has not changed, no.
